### PR TITLE
fix(audit): resolve more audit findings

### PR DESCRIPTION
**Pull request by MaxMustermann2**
fix: GWT-01
fix: TOW-02
fix: ENO-01

## Summary by CodeRabbit

- **New Features**
	- Enhanced multisignature verification process for improved consistency and gas handling.
	- Integrated authority-based permission management across various components, ensuring only valid authorities can update parameters on the mainnet.

- **Bug Fixes**
	- Restructured reward claiming methods to improve clarity and error handling for unsupported features.

- **Documentation**
	- Updated comments for better clarity on logic and functionality changes.

### CODE REVIEW

**Review by trestinlsd — APPROVED** [review]

**Review by mikebraver — APPROVED** [review]

**Comment by MaxMustermann2** [comment]
Refer to #220 for TOW-02

File: x/slash/keeper/keeper.go [hunk 1]
```diff
@@ -21,17 +21,21 @@ type Keeper struct {
...
 	cdc codec.BinaryCodec,
 	storeKey storetypes.StoreKey,
 	assetsKeeper keeper.Keeper,
+	authority string,
 ) Keeper {
 	return Keeper{
 		cdc:          cdc,
 		storeKey:     storeKey,
 		assetsKeeper: assetsKeeper,
+		authority:    authority,
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply x/slash/keeper/keeper.go]: These checks are added in each such keeper.

File: x/oracle/keeper/msg_server_update_params.go [hunk 2]
```diff
@@ -3,19 +3,22 @@ package keeper
...
-	//	if ms.authority != msg.Authority {
-	//		return nil, govtypes.ErrInvalidSigner.Wrapf("invalid authority; expected %s, got %s", ms.authority, msg.Authority)
-	//	}
-
+	if utils.IsMainnet(ctx.ChainID()) && ms.Keeper.authority != msg.Authority {
+		return nil, govtypes.ErrInvalidSigner.Wrapf(
+			"invalid authority; expected %s, got %s",
+			ms.Keeper.authority, msg.Authority,
+		)
+	}
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply x/oracle/keeper/msg_server_update_params.go:21]: Thanks, added

File: x/reward/keeper/claim_reward_test.go [hunk 3]
```diff
@@ -24,24 +24,26 @@ func (suite *RewardTestSuite) TestClaimWithdrawRequest() {
...
-	suite.ErrorContains(err, rewardtype.ErrRewardAssetNotExist.Error())
+	// suite.ErrorContains(err, rewardtype.ErrRewardAssetNotExist.Error())
+	suite.ErrorContains(err, rewardtype.ErrNotSupportYet.Error())
 
 	// test the normal case
 	event.AssetsAddress = usdtAddress[:]
 	err = suite.App.RewardKeeper.RewardForWithdraw(suite.Ctx, event)
-	suite.NoError(err)
+	// suite.NoError(err)
+	suite.ErrorContains(err, rewardtype.ErrNotSupportYet.Error())
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply x/reward/keeper/claim_reward_test.go:34]: Deprecated, will be later removed

File: x/reward/keeper/claim_reward_test.go [hunk 4]
```diff
@@ -24,24 +24,26 @@
...
-		PendingUndelegationAmount: sdkmath.NewInt(0),
-	}, *info)
+	// stakerID, assetID := types.GetStakerIDAndAssetID(event.ClientChainLzID, event.WithdrawRewardAddress, event.AssetsAddress)
+	// info, err := suite.App.AssetsKeeper.GetStakerSpecifiedAssetInfo(suite.Ctx, stakerID, assetID)
+	// suite.NoError(err)
+	// suite.Equal(types.StakerAssetInfo{
+	// 	TotalDepositAmount:        sdkmath.NewInt(10),
+	// 	WithdrawableAmount:        sdkmath.NewInt(10),
+	// 	PendingUndelegationAmount: sdkmath.NewInt(0),
+	// }, *info)
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply x/reward/keeper/claim_reward_test.go:44]: Deprecated, will be later removed

File: x/reward/keeper/claim_reward_test.go [hunk 5]
```diff
@@ -24,24 +24,26 @@
...
 
-	assetInfo, err := suite.App.AssetsKeeper.GetStakingAssetInfo(suite.Ctx, assetID)
-	suite.NoError(err)
-	suite.Equal(sdkmath.NewInt(10).Add(assets[0].StakingTotalAmount), assetInfo.StakingTotalAmount)
+	// assetInfo, err := suite.App.AssetsKeeper.GetStakingAssetInfo(suite.Ctx, assetID)
+	// suite.NoError(err)
+	// suite.Equal(sdkmath.NewInt(10).Add(assets[0].StakingTotalAmount), assetInfo.StakingTotalAmount)
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply x/reward/keeper/claim_reward_test.go:48]: Deprecated, will be later removed

File: testutil/keeper/oracle.go [hunk 6]
```diff
@@ -49,6 +51,7 @@ func OracleKeeper(t testing.TB) (*keeper.Keeper, sdk.Context) {
 		dogfoodkeeper.Keeper{},
 		delegationkeeper.Keeper{},
 		assetskeeper.Keeper{},
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply testutil/keeper/oracle.go:54]: Already done

File: precompiles/reward/reward_test.go [hunk 7]
```diff
@@ -83,7 +83,9 @@ func (s *RewardPrecompileTestSuite) TestRunRewardThroughClientChain() {
 		s.Require().NoError(err, "failed to pack input")
 		return s.Address, input
 	}
-	successRet, err := s.precompile.Methods[reward.MethodReward].Outputs.Pack(true, new(big.Int).Add(depositAmount, withdrawAmount))
+	// successRet, err := s.precompile.Methods[reward.MethodReward].Outputs.Pack(true, new(big.Int).Add(depositAmount, withdrawAmount))
+	// TODO: reward precompile is disabled, so it always errors and returns fail
+	successRet, err := s.precompile.Methods[reward.MethodReward].Outputs.Pack(false, new(big.Int))
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply precompiles/reward/reward_test.go:88]: Deprecated, will be later removed

File: x/exomint/keeper/msg_server.go [hunk 8]
```diff
@@ -27,13 +28,12 @@ func (k Keeper) UpdateParams(
...
-	// 		"invalid authority; expected %s, got %s",
-	// 		k.authority, msg.Authority,
-	// 	)
-	// }
+	if utils.IsMainnet(c.ChainID()) && k.authority != msg.Authority {
+		return nil, govtypes.ErrInvalidSigner.Wrapf(
+			"invalid authority; expected %s, got %s",
+			k.authority, msg.Authority,
+		)
+	}
```
[the remark this answers is not in the document]
Comment (MaxMustermann2) [reply x/exomint/keeper/msg_server.go:36]: Intentionally added that condition.

**Review by bwhour — APPROVED** [review]
looks good.

**Pull request merged**